tiri/src/lib.rs

```diff
@@ -11,28 +11,135 @@
+    BinOp, Constant, ConstantInt, Guard, Local, Operand, Place, PlaceBase, PlaceProjection, Rvalue,
+    SignedInt, Statement, TirOp, TirTrace, UnsignedInt,
+};
 
-/// Storage space for one local variable.
-/// FIXME: Not yet populated.
-struct Local {}
+/// The number of Tir statements to print either side of the PC when dumping the interpreter state.
+#[allow(dead_code)]
+const DUMP_CONTEXT: usize = 4;
+
+/// Each local variable in the interpreter maps to one `Value`.
+/// This pretty much emulates Miri's `Operand` type.
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Value {
+    Immediate(Immediate),
+    _Indirect, // FIXME Not implemented.
+}
+
+impl Value {
+    /// A helper to create a raw scalar value of the given value and size.
+    #[cfg(test)]
+    fn raw_scalar(data: u128, size: u8) -> Self {
+        Value::Immediate(Immediate::Scalar(Scalar::Raw { data, size }))
+    }
+
+    /// A helper to cast a value to a Boolean, if appropriate.
+    fn as_bool(&self) -> bool {
+        match self {
+            Value::Immediate(i) => i.as_bool(),
+            _ => panic!("Invalid cast from Value to bool"),
+        }
+    }
+
+    /// A helper to get the u128 value from inside a raw immediate.
+    fn as_u128(&self) -> u128 {
+        match self {
+            Value::Immediate(Immediate::Scalar(Scalar::Raw { data, .. })) => *data,
+            _ => panic!("Invalid cast from Value to u128"),
+        }
+    }
+}
+
+impl From<bool> for Value {
+    fn from(b: bool) -> Self {
+        Self::Immediate(Immediate::from(b))
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Immediate {
+    Scalar(Scalar),
+    ScalarPair(Scalar, Scalar),
+}
+
+impl From<bool> for Immediate {
+    fn from(b: bool) -> Self {
+        Self::Scalar(Scalar::from(b))
+    }
+}
+
+impl Immediate {
+    fn as_bool(&self) -> bool {
+        match self {
+            // In the case of a `ScalarPair` we assume we are casting to bool to inspect the result
+            // of a checked operation, hence we cast the second element.
+            Immediate::Scalar(s) | Immediate::ScalarPair(_, s) => s.as_bool(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+enum Scalar {
+    Raw { data: u128, size: u8 },
+    _Ptr, // FIXME Not implemented.
+}
+
+impl From<bool> for Scalar {
+    fn from(b: bool) -> Self {
+        if b {
+            Scalar::Raw { data: 1, size: 1 }
+        } else {
+            Scalar::Raw { data: 0, size: 1 }
+        }
+    }
+}
+
+impl Scalar {
+    fn as_bool(&self) -> bool {
+        let data = match self {
+            Scalar::Raw { data, .. } => data,
+            Scalar::_Ptr => panic!("Invalid cast from Scalar to bool"),
+        };
+        *data != 0
+    }
+}
 
 /// Mutable interpreter state.
-struct InterpState {
+pub struct InterpState {
     /// The next position in the trace to interpret.
-    trace_pos: usize,
+    pc: usize,
     /// Local variable store.
-    _locals: Vec<Local>,
+    locals: HashMap<u32, Value>,
```

Comment:
MaybEme use HashMap<Local, Value>, or even rustc_index::IndexMap<Local, Option<Value>> is most locals are expected to be initialized.

Reply:
Interesting. What are they initialised to before their first store?

Reply:
```diff
-rustc_index::IndexMap<Local, Option>
+rustc_index::IndexVec<Local, Option<Value>>`
```

> What are they initialised to before their first store?

You will need to add elements to it yourself, but using [`IndexVec::from_elem_n`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_index/vec/struct.IndexVec.html#method.from_elem_n) you can easily create an `IndexVec` with multiple identical elements, like `None`.